### PR TITLE
[Ready] F5-TTS Training Recipe for WenetSpeech4TTS

### PR DESCRIPTION
**Pull request by yuekaizhang**
Added F5-TTS Icefall style training codes based on https://github.com/SWivid/F5-TTS.

Training logs and results: https://huggingface.co/yuekai/f5-tts-small-wenetspeech4tts-basic
| Model | Seed-TTS test_zh WER| Comment |
|-------|---------------------------|------------|
|vall-e 150M | 4.33 | Icefall codebase reproduced, wenetspeech basic 7k hours |
| **f5-small 150M 32 steps** | 2.42 | Icefall codebase reproduced, wenetspeech basic 7k hours |
| f5-base  300M 32 steps | [1.52](https://github.com/SWivid/F5-TTS/issues/463#issuecomment-2475307364) | F5-TTS pretrained, Emilia zh&en 100k hours |

Samples:

https://github.com/user-attachments/assets/77393c61-c744-4b6c-b0bb-7c3fa1d27878

https://github.com/user-attachments/assets/72c4a64d-39e3-4cba-8a92-96ccbdfcffd3

### CODE REVIEW

**Comment by yfyeung** [comment]
Nice work

**Comment by JinZr** [comment]
ok! I’ll look into it this weekend

Best Regards
Jin


Yuekai Zhang ***@***.***>于2025年1月20日 周一17:56写道：

File: egs/wenetspeech4tts/TTS/f5-tts/vocab.txt
Comment (yfyeung) [inline egs/wenetspeech4tts/TTS/f5-tts/vocab.txt:1]: Could you provide instructions on how to generate this file? Thanks.

File: egs/wenetspeech4tts/TTS/prepare.sh [hunk 1]
```diff
@@ -98,3 +98,44 @@ if [ $stage -le 4 ] && [ $stop_stage -ge 4 ]; then
...
+    touch data/fbank/.${prefix}.done
+  fi
+fi
+
+if [ $stage -le 6 ] && [ $stop_stage -ge 6 ]; then
+  log "Stage 7: Split the ${prefix} cuts into train, valid and test sets (used by ./f5-tts)"
+  if [ ! -f data/fbank/${prefix}_cuts_${subset}.jsonl.gz ]; then
+    echo "Combining ${prefix} cuts"
+    pieces=$(find data/fbank/ -name "${prefix}_cuts_${subset}.*.jsonl.gz")
+    # lhotse combine $pieces data/fbank/${prefix}_cuts_${subset}.jsonl.gz
```
Comment (JinZr) [inline egs/wenetspeech4tts/TTS/prepare.sh:118]: the ''lhotse combine'' command is commented out, shouldn't this be executed after pieces are located to ensure proper integration?

File: egs/wenetspeech4tts/TTS/f5-tts/speech_synthesis.py
Comment (JinZr) [inline egs/wenetspeech4tts/TTS/f5-tts/speech_synthesis.py:1]: seems like this class provides the exact same functionality as the speech_synthesis dataset in lhotse, maybe we could try to use the lhotse built-in one to replace this

**Review by JinZr — CHANGES_REQUESTED** [review]

**Comment by JinZr** [comment]
hi! i just finished reviewing this pr and it looks great!

i've left a few comments in it, please take a look when you have a moment

thank you!

File: egs/wenetspeech4tts/TTS/prepare.sh (hunk 1 above)
Comment (yuekaizhang) [reply egs/wenetspeech4tts/TTS/prepare.sh:118]: Fixed.

File: egs/wenetspeech4tts/TTS/f5-tts/speech_synthesis.py
Comment (yuekaizhang) [reply egs/wenetspeech4tts/TTS/f5-tts/speech_synthesis.py:1]: The speech_synthesis class in Lhotse restricts the audio file to be specified and must exist.  See https://github.com/lhotse-speech/lhotse/blob/master/lhotse/dataset/speech_synthesis.py#L74

To use pre-extracted features from https://huggingface.co/datasets/yuekai/wenetspeech4tts_lhotse_manifest/tree/main/fbank without having to download the wav files (the former is only over 100 GB, while all wav files are nearly 1 TB), I used this custom file to comment out some unnecessary items, such as audio.

Comment (yuekaizhang) [reply egs/wenetspeech4tts/TTS/f5-tts/speech_synthesis.py:1]: At the same time, in the future, there may be further expansion of the existing custom speech_synthesis file, such as adding the key speech_semantic tokens; this separate file would be more flexible.

**Comment by JinZr** [comment]
i make a few commits to fix the flake8 and black formatting err, and i think this pr is good to go now.

**Review by JinZr — APPROVED** [review]
